homeassistant/components/knx/light.py

```diff
@@ -137,83 +133,77 @@ def max_mireds(self) -> int:
-        hs_color = kwargs.get(ATTR_HS_COLOR, self.hs_color)
-        white_value = kwargs.get(ATTR_WHITE_VALUE, self.white_value)
-        mireds = kwargs.get(ATTR_COLOR_TEMP, self.color_temp)
-
-        update_brightness = ATTR_BRIGHTNESS in kwargs
-        update_color = ATTR_HS_COLOR in kwargs
-        update_white_value = ATTR_WHITE_VALUE in kwargs
-        update_color_temp = ATTR_COLOR_TEMP in kwargs
-
-        # avoid conflicting changes and weird effects
-        if not (
-            self.is_on
-            or update_brightness
-            or update_color
-            or update_white_value
-            or update_color_temp
+        # ignore arguments if not supported to fall back to set_on()
+        brightness = (
+            kwargs.get(ATTR_BRIGHTNESS) if self._device.supports_brightness else None
+        )
+        mireds = (
+            kwargs.get(ATTR_COLOR_TEMP)
+            if self.color_mode == COLOR_MODE_COLOR_TEMP
+            else None
+        )
+        rgb = kwargs.get(ATTR_RGB_COLOR) if self.color_mode == COLOR_MODE_RGB else None
+        rgbw = (
+            kwargs.get(ATTR_RGBW_COLOR) if self.color_mode == COLOR_MODE_RGBW else None
+        )
+
+        if (
+            not self.is_on
+            and brightness is None
+            and mireds is None
+            and rgb is None
+            and rgbw is None
         ):
             await self._device.set_on()
+            return
 
-        if self._device.supports_brightness and (
-            update_brightness and not update_color
-        ):
-            # if we don't need to update the color, try updating brightness
-            # directly if supported; don't do it if color also has to be
-            # changed, as RGB color implicitly sets the brightness as well
-            await self._device.set_brightness(brightness)
-        elif (self._device.supports_rgbw or self._device.supports_color) and (
-            update_brightness or update_color or update_white_value
-        ):
-            # change RGB color, white value (if supported), and brightness
-            # if brightness or hs_color was not yet set use the default value
-            # to calculate RGB from as a fallback
-            if brightness is None:
-                brightness = DEFAULT_BRIGHTNESS
-            if hs_color is None:
-                hs_color = DEFAULT_COLOR
-            if white_value is None and self._device.supports_rgbw:
-                white_value = DEFAULT_WHITE_VALUE
-            hsv_color = hs_color + (brightness * 100 / 255,)
-            rgb = color_util.color_hsv_to_RGB(*hsv_color)
-            await self._device.set_color(rgb, white_value)
-
-        if update_color_temp:
+        # return after RGB(W) color has changed as it implicitly sets the brightness
+        if rgbw is not None:
+            await self._device.set_color(rgbw[:3], rgbw[3])
+            return
+        if rgb is not None:
+            await self._device.set_color(rgb, None)
+            return
```

Comment:
Note that the light should be able to handle both a `brightness` parameter and an `rgb_color` or `rgbw_color` parameter.
Does the KNX api not have a master brightness feature, it expects brightness information to be set via `set_color` and read from `current_color`?

Reply:
KNX doesn't have a real API - at least its not specified how device features are controlled. And there are hundreds of different devices from lots of different manufacturers implementing things slightly different. So we try to cover every possible option 🤷‍♂️ (I've seen individual % values without binary switch for RGB, individual HSVW values, combined RGB and RGBW values, brightness and kelvin values, etc. Some reoport current status, some don't. Its a mess).

Reply:
I don't understand what this means
> The light can be dimmed and its color is present in the state as an (r, g, b) tuple (not normalized for brightness).
Is it that a 50% dimmed red light receives a turn_on with 
```python
{ "rgb_color": (255, 0, 0),
  "brightness": 128 }
```
or would it receive 
```python
{ "rgb_color": (128, 0, 0),
  "brightness": 128 }
```
?

Reply:
The `turn_on` only receives what's in the service call, so this would be expected to set the light's red channel to 100%:
`{ "rgb_color": (255, 0, 0), "brightness": 255 }`

Both of these would be expected to set the light's red channel to 50%:
`{ "rgb_color": (128, 0, 0), "brightness": 255 }`
`{ "rgb_color": (255, 0, 0), "brightness": 128 }`

Whereas this would be expected to set the light's red channel to 25%:
`{ "rgb_color": (128, 0, 0), "brightness": 128 }`

If the light is first set to its red channel at 50%:
`{ "rgb_color": (128, 0, 0), "brightness": 255 }`

This would be expected to dim the red channel to 25%:
`{ "brightness": 128 }`

And this would be expected to set it back to 50%:
`{ "brightness": 255 }`